### ISSUE TITLE
Form element is discarded due to LOWERCASE content encoding

### ISSUE THREAD

**Issue by peskybp** [jhy/jsoup#877]
The following HTML parses incorrectly ONLY when using a Jsoup.parse method that routes through the DataUtil class (where character set is determined from META comparisons). Parsing it from a string works just fine.

[file.html.txt](https://github.com/jhy/jsoup/files/971788/file.html.txt)
```
<!doctype html>
<html lang=en>
<meta charset=utf-8>
<title>FormData: has</title>
<script src="../../resources/testharness.js"></script>
<script src="../../resources/testharnessreport.js"></script>
    <link rel="help" href="https://xhr.spec.whatwg.org/#dom-formdata-get" />
    <link rel="help" href="https://xhr.spec.whatwg.org/#dom-formdata-getall" />
<div id="log"></div>
<form id="form">
    <input type="hidden" name="key" value="value1">
    <input type="hidden" name="key" value="value2">
</form>
<form id="empty-form" />
```

The form with id="form" gets discarded, the result of processing is the following:

```
<!doctype html>
<html lang="en">
 <head>
  <meta charset="utf-8">
  <title>FormData: has</title>
  <script src="../../resources/testharness.js"></script>
  <script src="../../resources/testharnessreport.js"></script>
  <link rel="help" href="https://xhr.spec.whatwg.org/#dom-formdata-get">
  <link rel="help" href="https://xhr.spec.whatwg.org/#dom-formdata-getall">
 </head>
 <body>
  <div id="log"></div>
  <input type="hidden" name="key" value="value1">
  <input type="hidden" name="key" value="value2">
  <form id="empty-form">
  </form>
 </body>
</html>
```

If the charset is changed from utf-8 to UTF-8 then it works out as intended.  So, definitely something to be done there about case insensitive comparisons, but that certainly isn't the end of the issue.

As far as I can tell, the Document is already incorrect BEFORE setting the doc.outputSettings().charset(..) in DataUtil.parseByteData(..). Wondering if there is perhaps some state that exists on the TreeBuilder or Parser instance overall.

**Comment by peskybp** [comment]
```
<!doctype html>
<html lang=en>
<meta charset=utf-8>
<title>FormData: has</title>
<script src="../../resources/testharness.js"></script>
<script src="../../resources/testharnessreport.js"></script>
    <link rel="help" href="https://xhr.spec.whatwg.org/#dom-formdata-get" />
    <link rel="help" href="https://xhr.spec.whatwg.org/#dom-formdata-getall" />
<div id="log"></div>
<form id="form">
    <input type="hidden" name="key" value="value1">
    <input type="hidden" name="key" value="value2">
</form>
<form id="empty-form" />

**Comment by jhy** [comment]
Thanks for the report and the detail!

That's fixed now with 1203bd459fe137e78a2a6dc2969f3c4995b726ce (which describes the issue in a bit more detail, but in essence, the HTML parser was being reused without being reset correctly, and DataUtil was (incorrectly) deciding it needed to redecode the doc).
